sdk/java/pulumi/src/main/java/com/pulumi/provider/internal/properties/PropertyValueSerializer.java

```diff
@@ -366,4 +371,211 @@ private static String propertyName(Field field, Class<?> declaringType) {
+                    Map.Entry::getValue,
+                    (a, b) -> a,
+                    HashMap::new
+                )));
+    }
+
+    /**
+     * Converts a component resource object into a map of property values.
+     * This method examines the fields of the component resource marked with the {@link Export} annotation
+     * and serializes their values into {@link PropertyValue} objects.
+     *
+     * @param component The component resource object to convert
+     * @return A map where keys are property names and values are serialized {@link PropertyValue} objects
+     * @throws PropertySerializationException if there is an error accessing field values
+     */
+    public static Map<String, PropertyValue> stateFromComponentResource(Object component) {
+        return stateFromComponentResourceAsync(component).join();
+    }
+
+    private static CompletableFuture<PropertyValue> serialize(Object value, String... path) {
+        if (value == null) {
+            return CompletableFuture.completedFuture(PropertyValue.NULL);
+        }
+
+        if (value instanceof Output<?>) {
+            return serializeOutput((Output<?>) value, path);
+        }
+        return CompletableFuture.completedFuture(serializeValue(value, path));
+    }
+
+    private static CompletableFuture<PropertyValue> serializeOutput(Output<?> output, String... path) {
+        return Internal.of(output).getDataAsync().thenCompose(data -> {
+            if (!data.isKnown()) {
+                return CompletableFuture.completedFuture(PropertyValue.COMPUTED);
+            }
+
+            return serialize(data.getValueNullable()).thenApply(element -> {
+                var dependencies = data.getResources().stream()
+                    .map(resource -> Internal.of(resource.urn()))
+                    .map(OutputInternal::getDataAsync)
+                    .map(urnFuture -> urnFuture
+                        .thenApply(OutputData::getValueNullable)
+                        .join())
+                    .collect(Collectors.toSet());
+
+                PropertyValue outputValue;
+                if (dependencies.isEmpty()) {
+                    outputValue = element != null ? element : PropertyValue.COMPUTED;
+                } else {
+                    outputValue = PropertyValue.of(
+                        new PropertyValue.OutputReference(element, dependencies));
+                }
+
+                return data.isSecret() ? PropertyValue.ofSecret(outputValue) : outputValue;
+            });
+        });
+    }
+
+    private static PropertyValue serializeValue(Object value, String... path) {
+        if (value == null) {
+            return PropertyValue.NULL;
+        }
+
+        // Handle primitive types and their wrappers
+        if (value instanceof String) {
+            return PropertyValue.of((String) value);
+        }
+
+        if (value instanceof Integer) {
+            return PropertyValue.of(((Integer) value).doubleValue());
```

Comment:
Shouldn't this be 

```suggestion
            return PropertyValue.of(((Integer) value).intValue());
```

?  The property value type will still be "Number", but the value object will be of type Integer if we use `.intValue()`, which I think makes a difference when we deserialize the value.  My java knowledge is pretty outdated though, so I might be missing something.

Reply:
I think the difference would be [here](https://github.com/pulumi/pulumi-java/blob/main/sdk/java/pulumi/src/main/java/com/pulumi/provider/internal/properties/PropertyValue.java#L134-L136), where if we pass in an int value we store an `Integer` object here, which you can get back out [here](https://github.com/pulumi/pulumi-java/blob/main/sdk/java/pulumi/src/main/java/com/pulumi/provider/internal/properties/PropertyValue.java#L171).  Though for [getNumberValue](https://github.com/pulumi/pulumi-java/blob/main/sdk/java/pulumi/src/main/java/com/pulumi/provider/internal/properties/PropertyValue.java#L203) we always return doubles, so maybe this is right as it is.

Reply:
Though maybe then there's a bug in `PropertyValue of(int value)`, and it should be `return new PropertyValue(ValueType.NUMBER, Double.valueOf(value));` instead :thinking:

Reply:
Good point. PropertyValue of(int value) is actually currently never used, so I'll just remove it.

Reply:
Sounds good to me!